compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,565 @@
+    ///     fn default() -> Foo {
+    ///         Foo { x: 100, y: NonDefault }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Manually writing a `Default` implementation for a type that has
+    /// default field values runs the risk of diverging behavior between
+    /// `Type { .. }` and `<Type as Default>::default()`, which would be a
+    /// foot-gun for users of that type that would expect these to be
+    /// equivalent. If `Default` can't be derived due to some fields not
+    /// having a `Default` implementation, we encourage the use of `..` for
+    /// the fields that do have a default field value.
+    pub DEFAULT_OVERRIDES_DEFAULT_FIELDS,
+    Warn,
+    "detect `Default` impl that should use the type's default field values",
+    @feature_gate = default_field_values;
+}
+
+#[derive(Default)]
+pub(crate) struct DefaultCouldBeDerived {
+    data: Option<Data>,
+}
+
+struct Data {
+    type_def_id: DefId,
+    parent: DefId,
+}
+
+impl_lint_pass!(DefaultCouldBeDerived => [DEFAULT_COULD_BE_DERIVED, DEFAULT_OVERRIDES_DEFAULT_FIELDS]);
+
+impl<'tcx> LateLintPass<'tcx> for DefaultCouldBeDerived {
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &hir::ImplItem<'_>) {
+        let Some(default_def_id) = cx.tcx.get_diagnostic_item(sym::Default) else { return };
+        let assoc = cx.tcx.associated_item(impl_item.owner_id);
+        let parent = assoc.container_id(cx.tcx);
+        if cx.tcx.has_attr(parent, sym::automatically_derived) {
+            return;
+        }
+        let Some(trait_ref) = cx.tcx.impl_trait_ref(parent) else { return };
+        let trait_ref = trait_ref.instantiate_identity();
+        if trait_ref.def_id != default_def_id {
+            return;
+        }
+        let ty = trait_ref.self_ty();
+        let ty::Adt(def, _) = ty.kind() else { return };
+
+        // We have a manually written definition of a `<Type as Default>::default()`. We store the
+        // necessary metadata for further analysis of its body in `check_body`.
+        self.data = Some(Data { type_def_id: def.did(), parent });
+    }
+
+    fn check_impl_item_post(&mut self, _cx: &LateContext<'_>, _impl_item: &hir::ImplItem<'_>) {
+        // Clean up.
+        self.data = None;
+    }
+
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &hir::Body<'tcx>) {
+        // We perform this logic here instead of in `check_impl_item` so that we have unconditional
+        // permission to call `cx.typeck_results()` (because we're within a body).
+        let Some(Data { type_def_id, parent }) = self.data else {
+            return;
+        };
+        // FIXME: evaluate bodies with statements and evaluate bindings to see if they would be
+        // derivable.
+        let hir::ExprKind::Block(hir::Block { stmts: _, expr: Some(expr), .. }, None) =
+            body.value.kind
+        else {
+            return;
+        };
```

Comment:
I can remove all of these checks to a single method, but then I would end up returning a big tuple of unrelated fields. I leave it to you whether it is preferable to have shorter methods with descriptive names for these steps, or more linear logic with comments.